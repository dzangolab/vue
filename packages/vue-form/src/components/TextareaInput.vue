<template>
  <div :class="`field ${name}`">
    <label v-if="label" :for="name">
      {{ label }}
    </label>
    <Field
      v-slot="{ field, meta }"
      :model-value="modelValue"
      :name="name"
      :rules="fieldSchema"
      @input="onInput"
    >
      <textarea
        v-bind="field"
        :id="`input-field-${name}`"
        :class="{
          invalid: meta.touched && !meta.valid,
          valid: meta.dirty && meta.valid,
        }"
        :disabled="disabled"
        :placeholder="placeholder"
        tabindex="0"
        type="textarea"
      />
      <ErrorMessage :name="name" />
    </Field>
  </div>
</template>

<script lang="ts">
export default {
  name: "TextareaInput",
};
</script>

<script setup lang="ts">
import { toFieldValidator } from "@vee-validate/zod";
import { ErrorMessage, Field } from "vee-validate";
import { z } from "zod";

import { textSchema } from "../schemas";

import type { TextErrorMessages, IsTextOptions } from "../types";
import type { PropType } from "vue";

const props = defineProps({
  disabled: {
    default: false,
    type: Boolean,
  },
  errorMessages: {
    default: () => {
      return {
        invalid: "Please provide a valid input data",
        required: "The field is required",
      };
    },
    required: false,
    type: Object as PropType<TextErrorMessages>,
  },
  label: {
    default: "",
    required: false,
    type: String as PropType<string>,
  },
  modelValue: {
    default: "",
    type: String as PropType<string | number | string[] | undefined>,
  },
  name: {
    default: "text",
    required: false,
    type: String as PropType<string>,
  },
  options: {
    default: () => {
      return {};
    },
    required: false,
    type: Object as PropType<IsTextOptions>,
  },
  placeholder: {
    default: "",
    type: String,
  },
  schema: {
    default: () => {
      return {};
    },
    required: false,
    type: Object as PropType<z.ZodType<string | number>>,
  },
});

const emit = defineEmits(["update:modelValue"]);

const fieldSchema = toFieldValidator(
  Object.keys(props.schema).length
    ? props.schema
    : textSchema(props.errorMessages, props.options)
);

const onInput = (event: Event) => {
  const value = (event.target as HTMLInputElement).value;

  emit("update:modelValue", value);
};
</script>

<style lang="css">
.field textarea {
  --_input-field-textarea-height: var(--textarea-height, 8rem);
  --_input-field-textarea-resize: var(--textarea-resize, none);
  --_input-field-textarea-width: var(--textarea-width, 100%);

  height: var(--_input-field-textarea-height);
  resize: var(--_input-field-textarea-resize);
  width: var(--_input-field-textarea-width);
}
</style>
