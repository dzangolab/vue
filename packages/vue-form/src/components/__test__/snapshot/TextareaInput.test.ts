import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import TextareaInput from "../../TextareaInput.vue";

describe("TextareaInput", () => {
  it("matches snapshot", () => {
    const wrapper = mount(TextareaInput, {
      props: {
        label: "Textarea input",
        placeholder: "Placeholder",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
