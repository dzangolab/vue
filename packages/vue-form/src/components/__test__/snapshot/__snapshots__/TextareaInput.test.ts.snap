// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextareaInput > matches snapshot 1`] = `
<div
  class="field text"
  data-v-f3edb0b1=""
>
  <label
    data-v-f3edb0b1=""
    for="text"
  >
    Textarea input
  </label>
  
  <textarea
    class=""
    data-v-f3edb0b1=""
    id="input-field-text"
    name="text"
    placeholder="Placeholder"
    tabindex="0"
    type="textarea"
  />
  <!---->
  
</div>
`;
